### PR TITLE
Peel-off AbstractChannelHandlerContext's ChannelHandlers call-sites

### PR DESCRIPTION
**Pull request by franz1981**
Motivation:
AbstractChannelHandlerContext invoke* methods often fall-off to itable_stub due to interface megamorphism, causing frequent (and mandatory) handlers to not benefit from inlining.

Modification:
Add instanceof checks to peel-off AbstractChannelHandlerContext call-sites for the most frequent (and possibly, mandatory) ChannelHandler's concrete types, to both reduce the arity of the remaining call-site, to improve likelihood of mono/bi-morphism on it, while making possible inlining for the checked ones.

Result:
Faster and more predictable (without itable_stub) ChannelHandlers methods calls propagation.

### CODE REVIEW

**Comment by franz1981** [comment]
I've collected call-sites type profiling on different Techempower benchmarks using both Vert-x and Quarkus, but I'm sure some of the checks I've performed makes sense for other users/common Netty users (to be defined): please @normanmaurer feel free to suggest different ones otherwise (and/or fix the ones I've found).

Probably Netty 5 requires a similar treatment due to how JIT works (https://shipilev.net/blog/2015/black-magic-method-dispatch/ for more info; ancient but still very good) 

Furthermore: given that we have concrete types built-in on Netty that we care about, we can even decide to enrich the existing list with more, saving any virtual calls and favour inlining (whatever possible) to happen for them.

@njhill did you see this already?

**Comment by franz1981** [comment]
@trustin did you noticed anything about it on Armenia as well?  Suggestions are wellcome :D

**Comment by normanmaurer** [comment]
@franz1981 hmm.. But wouldn't this hurt all the other "handlers" as it will always do the instanceof checks and it will only work for the head and tail ?

**Comment by franz1981** [comment]
@normanmaurer 
Right, if we got itable_stub it means that in a callchain  "someone" called a method of the context that didn't got inlined, meaning that use a (shared) compiled version of it, sharing type profiling information of different handlers called into it.
Hitting the itable_stub on some handler method call means that JIT is going to use an optimized dispatch/jump table that still contains these checks (head/tail) because these are types seen with some probability at runtime.
The handle call would just not benefit from being inlined in the context body and the jump table requires jumping in a differente assembly location (that's not super icache friendly).

Thanks to your comment, I see that the real question is on the beginning of this same comment 

> "someone" called a method of the context that didn't got inlined
 
I'll investigate a but more

**Comment by normanmaurer** [comment]
> @normanmaurer Right, if we got itable_stub it means that in a callchain "someone" called a method of the context that didn't got inlined, meaning that use a (shared) compiled version of it, sharing type profiling information of different handlers called into it. Hitting the itable_stub on some handler method call means that JIT is going to use an optimized dispatch/jump table that still contains these checks (head/tail) because these are types seen with some probability at runtime. The handle call would just not benefit from being inlined in the context body and the jump table requires jumping in a differente assembly location (that's not super icache friendly).
> 
> Thanks to your comment, I see that the real question is on the beginning of this same comment
> 
> > "someone" called a method of the context that didn't got inlined
> 
> Why?
> 
> I'll investigate a bit more, but maybe you got some clue already here: the idea is to make a handler chain event/call to be more JIT friendly

So the question was really if this makes sense at all as sure it helps for the tail and head but imho it might hurt for the rest of the pipeline ass you basically do two instanceof checks that almost always will fail (these will only be true for two handlers).

**Comment by franz1981** [comment]
> So the question was really if this makes sense at all as sure it helps for the tail and head but imho it might hurt for the rest of the pipeline ass you basically do two instanceof checks that almost always will fail (these will only be true for two handlers).

This is going to be true only if intermediate contexts fireXXX methods are inlined: if inlined these checked will mostly fail (and the JIT can decide to move it elsewhere given that are "unlikely" to happen

if these methods are not inlined, but compiled, it means that *any* context will share the same fireXXX method with the same call-site type profiling data: it means that the handler call will likely become megamorphic and itable_stub will still perform the check vs head/tail types anyway (because these are seen types as well) but:
- moving into the assembly region that perform the itable_stub call dispatch
- not making the channel handlers (head/tail) calls to benefit from inlining 

In short: if not inlined, these checks will still be performed, but with less optimizations (re inlining and instruction-cache sympathy), if inlined they still be nicely predicted (and correctly placed by the JIT) because almost false for that assembly code

**Comment by doom369** [comment]
Can you please share the call sites you mentioned? Also, do you have a benchmark that shows the difference?

**Comment by franz1981** [comment]
@normanmaurer @doom369 
> Can you please share the call sites you mentioned?

image

> Also, do you have a benchmark that shows the difference?

Still WIP, same for compilation logs (the perf issue is low prio ATM), but the whole channel pipeline here doesn't have lot of handlers defined; it's a normal HTTP 1.1 pipeline :)

File: transport/src/main/java/io/netty/channel/AbstractChannelHandlerContext.java [hunk 1]
```diff
@@ -343,7 +371,12 @@ public void run() {
     private void invokeUserEventTriggered(Object event) {
         if (invokeHandler()) {
             try {
-                ((ChannelInboundHandler) handler()).userEventTriggered(this, event);
+                final ChannelHandler handler = handler();
+                if (handler instanceof DefaultChannelPipeline.TailContext) {
+                    ((DefaultChannelPipeline.TailContext) handler).userEventTriggered(null, event);
```
Comment (normanmaurer) [inline transport/src/main/java/io/netty/channel/AbstractChannelHandlerContext.java:376]:
```suggestion
                    ((DefaultChannelPipeline.TailContext) handler).userEventTriggered(this, event);
```

File: transport/src/main/java/io/netty/channel/AbstractChannelHandlerContext.java [hunk 2]
```diff
@@ -259,7 +280,14 @@ public void run() {
     private void invokeChannelInactive() {
         if (invokeHandler()) {
             try {
-                ((ChannelInboundHandler) handler()).channelInactive(this);
+                final ChannelHandler handler = handler();
+                if (handler instanceof DefaultChannelPipeline.HeadContext) {
+                    ((DefaultChannelPipeline.HeadContext) handler).channelInactive(this);
+                } else if (handler instanceof DefaultChannelPipeline.TailContext) {
+                    ((DefaultChannelPipeline.TailContext) handler).channelInactive(null);
```
Comment (normanmaurer) [inline transport/src/main/java/io/netty/channel/AbstractChannelHandlerContext.java:287]:
```suggestion
                    ((DefaultChannelPipeline.TailContext) handler).channelInactive(this);
```

File: transport/src/main/java/io/netty/channel/AbstractChannelHandlerContext.java [hunk 3]
```diff
@@ -163,7 +163,14 @@ public void run() {
     private void invokeChannelRegistered() {
         if (invokeHandler()) {
             try {
-                ((ChannelInboundHandler) handler()).channelRegistered(this);
+                final ChannelHandler handler = handler();
+                if (handler instanceof DefaultChannelPipeline.HeadContext) {
+                    ((DefaultChannelPipeline.HeadContext) handler).channelRegistered(this);
+                } else if (handler instanceof DefaultChannelPipeline.TailContext) {
+                    ((DefaultChannelPipeline.TailContext) handler).channelRegistered(null);
```
Comment (normanmaurer) [inline transport/src/main/java/io/netty/channel/AbstractChannelHandlerContext.java:170]:
```suggestion
                    ((DefaultChannelPipeline.TailContext) handler).channelRegistered(this);
```

File: transport/src/main/java/io/netty/channel/AbstractChannelHandlerContext.java [hunk 4]
```diff
@@ -195,7 +202,14 @@ public void run() {
     private void invokeChannelUnregistered() {
         if (invokeHandler()) {
             try {
-                ((ChannelInboundHandler) handler()).channelUnregistered(this);
+                final ChannelHandler handler = handler();
+                if (handler instanceof DefaultChannelPipeline.HeadContext) {
+                    ((DefaultChannelPipeline.HeadContext) handler).channelUnregistered(this);
+                } else if (handler instanceof DefaultChannelPipeline.TailContext) {
+                    ((DefaultChannelPipeline.TailContext) handler).channelUnregistered(null);
```
Comment (normanmaurer) [inline transport/src/main/java/io/netty/channel/AbstractChannelHandlerContext.java:209]:
```suggestion
                    ((DefaultChannelPipeline.TailContext) handler).channelUnregistered(this);
```

File: transport/src/main/java/io/netty/channel/AbstractChannelHandlerContext.java [hunk 5]
```diff
@@ -227,7 +241,14 @@ public void run() {
     private void invokeChannelActive() {
         if (invokeHandler()) {
             try {
-                ((ChannelInboundHandler) handler()).channelActive(this);
+                final ChannelHandler handler = handler();
+                if (handler instanceof DefaultChannelPipeline.HeadContext) {
+                    ((DefaultChannelPipeline.HeadContext) handler).channelActive(this);
+                } else  if (handler instanceof DefaultChannelPipeline.TailContext) {
+                    ((DefaultChannelPipeline.TailContext) handler).channelActive(null);
```
Comment (normanmaurer) [inline transport/src/main/java/io/netty/channel/AbstractChannelHandlerContext.java:248]:
```suggestion
                    ((DefaultChannelPipeline.TailContext) handler).channelActive(this);
```

File: transport/src/main/java/io/netty/channel/AbstractChannelHandlerContext.java [hunk 6]
```diff
@@ -407,7 +445,14 @@ static void invokeChannelReadComplete(final AbstractChannelHandlerContext next)
     private void invokeChannelReadComplete() {
         if (invokeHandler()) {
             try {
-                ((ChannelInboundHandler) handler()).channelReadComplete(this);
+                final ChannelHandler handler = handler();
+                if (handler instanceof DefaultChannelPipeline.HeadContext) {
+                    ((DefaultChannelPipeline.HeadContext) handler).channelReadComplete(this);
+                } else if (handler instanceof DefaultChannelPipeline.TailContext) {
+                    ((DefaultChannelPipeline.TailContext) handler).channelReadComplete(null);
```
Comment (normanmaurer) [inline transport/src/main/java/io/netty/channel/AbstractChannelHandlerContext.java:452]:
```suggestion
                    ((DefaultChannelPipeline.TailContext) handler).channelReadComplete(this);
```

File: transport/src/main/java/io/netty/channel/AbstractChannelHandlerContext.java [hunk 7]
```diff
@@ -438,7 +483,14 @@ static void invokeChannelWritabilityChanged(final AbstractChannelHandlerContext
     private void invokeChannelWritabilityChanged() {
         if (invokeHandler()) {
             try {
-                ((ChannelInboundHandler) handler()).channelWritabilityChanged(this);
+                final ChannelHandler handler = handler();
+                if (handler instanceof DefaultChannelPipeline.HeadContext) {
+                    ((DefaultChannelPipeline.HeadContext) handler).channelWritabilityChanged(this);
+                } else if (handler instanceof DefaultChannelPipeline.TailContext) {
+                    ((DefaultChannelPipeline.TailContext) handler).channelWritabilityChanged(null);
```
Comment (normanmaurer) [inline transport/src/main/java/io/netty/channel/AbstractChannelHandlerContext.java:490]:
```suggestion
                    ((DefaultChannelPipeline.TailContext) handler).channelWritabilityChanged(this);
```

File: transport/src/main/java/io/netty/channel/AbstractChannelHandlerContext.java [hunk 8]
```diff
@@ -714,7 +766,12 @@ void invokeWrite(Object msg, ChannelPromise promise) {
 
     private void invokeWrite0(Object msg, ChannelPromise promise) {
         try {
-            ((ChannelOutboundHandler) handler()).write(this, msg, promise);
+            final ChannelHandler handler = handler();
+            if (handler instanceof DefaultChannelPipeline.HeadContext) {
+                ((DefaultChannelPipeline.HeadContext) handler).write(null, msg, promise);
```
Comment (normanmaurer) [inline transport/src/main/java/io/netty/channel/AbstractChannelHandlerContext.java:771]:
```suggestion
                ((DefaultChannelPipeline.HeadContext) handler).write(this, msg, promise);
```

File: transport/src/main/java/io/netty/channel/AbstractChannelHandlerContext.java [hunk 9]
```diff
@@ -747,7 +804,12 @@ private void invokeFlush() {
 
     private void invokeFlush0() {
         try {
-            ((ChannelOutboundHandler) handler()).flush(this);
+            final ChannelHandler handler = handler();
+            if (handler instanceof DefaultChannelPipeline.HeadContext) {
+                ((DefaultChannelPipeline.HeadContext) handler).flush(null);
```
Comment (normanmaurer) [inline transport/src/main/java/io/netty/channel/AbstractChannelHandlerContext.java:809]:
```suggestion
                ((DefaultChannelPipeline.HeadContext) handler).flush(this);
```

**Review by normanmaurer — CHANGES_REQUESTED** [review]

**Comment by franz1981** [comment]
@doom369 did you get similar figures in prod envs?

**Comment by doom369** [comment]
@franz1981 yes. I have seen that. However, the last time I did profiling - it was a JVM 11, 1 year ago, now we're on JVM 17 (and there were many changes since). What JVM version did you test against?

**Comment by franz1981** [comment]
@doom369 Recently having fun with how JDK treat megamorphism and inline caches and I can tell we won't get that much benefit by new versions on that (the version I've tested was both jdk 11 and the very latest LTS)

**Comment by chrisvest** [comment]
If the latest JITs can derive the same or similar optimisation from cleaner code, then I'm partial to having cleaner code and let people on older JVMs have marginally worse performance.

**Comment by franz1981** [comment]
Sorry @chrisvest catching up, I see my comment wasn't clear on

> I can tell we won't get that much benefit by new versions on that (the version I've tested was both jdk 11 and the very latest LTS)

I mean that new JDK version aren't still able to improve things here: we need a different approach, although I see that would make the code worse - let me try in our perf rig what kind of improvement I can get.
On JMH too
